source/languageDetection.py

```diff
@@ -0,0 +1,267 @@
+	allLanguages.sort()
+	languageDescriptions = []
+	if ignoreLanguages is None:
+		ignoreLanguages = {lang.languageID for lang in languagePriorityListSpec}
+	for language in allLanguages:  
+		if language in ignoreLanguages: 
+			continue
+		else:
+			desc=languageHandler.getLanguageDescription(language )
+			label="%s, %s"%(desc,language ) if desc else language 
+			languageDescriptions.append( LanguageDescription(languageID= language , description = label))
+	return languageDescriptions
+
+def getScriptCode(chr):
+	"""performs a binary search in scripCodes for unicode ranges
+	@param chr: character for which a script should be found
+	@type chr: string
+	@return: script code
+	@rtype: int"""
+	mStart = 0
+	mEnd = len(scriptCode)-1
+	characterUnicodeCode = ord(chr)
+	# Number should respect preferred language setting
+	# FullWidthNumber is in Common category, however, it indicates Japanese language context
+	if DIGIT_ZERO  <= characterUnicodeCode <= DIGIT_NINE:
+		return "Number"
+	elif FULLWIDTH_ZERO <= characterUnicodeCode <= FULLWIDTH_NINE: 
+		return "FullWidthNumber"
+	while( mEnd >= mStart ):
+		midPoint = (mStart + mEnd ) >> 1
+		if characterUnicodeCode < scriptCode[midPoint][0]: 
+			mEnd = midPoint -1
+		elif characterUnicodeCode > scriptCode[midPoint][1]: 
+			mStart = midPoint + 1
+		else:
+			return scriptCode[midPoint][2] 
+	return None
+
+def getLangID(scriptName ):
+	"""This function is the heart of determining which language is selected for a script
+	@param scriptName: the unicode name of the  script
+	@type scriptName: string
+	@return: It returns languageID for a script. it first checks whether there is a language for the script in the languagePriorityListSpec. If not, then it checks whether default language script is same as the script. At last it returns languageID for the script from scriptIDToLangID.
+	@rtype: string"""
+	# we are using loop during search to maintain priority
+	for priorityLanguage in  languagePriorityListSpec:
+		if scriptName in priorityLanguage.scriptID:
+			return priorityLanguage.languageID
+	#language not found in the languagePriorityListSpec, so check if default language can be applied for the script
+	if scriptName == getScriptIDFromLangID(languageHandler.getLanguage() ):
+		return  languageHandler.getLanguage()
+	# default language is not applicable for this script, so look up in the scriptIDToLangID
+	return scriptIDToLangID.get (scriptName )
+
+def getLanguageDescription(language ):
+	desc=languageHandler.getLanguageDescription(language )
+	label="%s, %s"%(desc,language ) if desc else language 
+	return label
+
+def getScriptIDFromLangID(langID ):
+	# Strip the dialect (if any).
+	return langIDToScriptID.get (langID.split("_", 1)[0])
+
+class ScriptChangeCommand(SpeechCommand):
+	"""A command to switch the script during script detection ."""
+
+	def __init__(self, scriptCode):
+		"""
+		@param scriptCode: the script identifier
+		@type scriptCode: int
+		"""
+		self.scriptCode =scriptCode 
```

Comment:
I'm a little bit worried about the name collision with the import on line 14: `from unicodeScriptData import scriptCode`. I don't think there is anywhere where this causes a bug. But I think it's a bit confusing, and isn't immediately obvious.

Reply:
removed the confusion by moving getScriptCode function from languageDetection.py to unicodeScriptData.py. Earlier I did not put this function in unicodeScriptData as the entire file was being generating by unicodeScriptPrep. now unicodeScriptPrep is generating unicodeScriptDataTemp from which the scriptRanges can be copied to unicodeScriptData. scriptRanges list was earlier known as scriptCode.